### PR TITLE
Rewrite of the AppGW module

### PR DESCRIPTION
**Pull request by rahalan**
# Change

New approach for the AppGW module. As it is very versatile, many properties are just arrays

[Network: ApplicationGateways](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.applicationgateways.yml)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

**Review by eriqua — COMMENTED** [review]
General question to the group: should we start updating the version.json for breaking changes or not yet?

File: arm/Microsoft.Network/applicationGateways/.parameters/parameters.json [hunk 1]
```diff
@@ -3,159 +3,358 @@
...
-        },
-        "vNetResourceGroup": {
-            "value": "validation-rg"
-        },
-        "frontendPrivateIpAddress": {
-            "value": "10.0.8.6"
-        },
-        "frontendPublicIpResourceId": {
-            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/publicIPAddresses/adp-<<namePrefix>>-az-pip-x-apgw"
+            "value": "sxx-az-apgw-x-001"
```
Comment (eriqua) [inline arm/Microsoft.Network/applicationGateways/.parameters/parameters.json:6]:
```suggestion
            "value": "<<namePrefix>>-az-apgw-x-001"
```

File: arm/Microsoft.Network/applicationGateways/.parameters/parameters.json [hunk 2]
```diff
@@ -3,159 +3,358 @@
...
         },
         "userAssignedIdentities": {
             "value": {
-                "/subscriptions/<<subscriptionId>>/resourcegroups/validation-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/adp-<<namePrefix>>-az-msi-x-001": {}
+                "/subscriptions/<<subscriptionId>>/resourcegroups/validation-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/adp-sxx-az-msi-x-001": {}
```
Comment (eriqua) [inline arm/Microsoft.Network/applicationGateways/.parameters/parameters.json:10]:
```suggestion
                "/subscriptions/<<subscriptionId>>/resourcegroups/validation-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/adp-<<namePrefix>>-az-msi-x-001": {}
```

File: arm/Microsoft.Network/applicationGateways/.parameters/parameters.json [hunk 3]
```diff
@@ -3,159 +3,358 @@
...
+        //     "enabled": true,
+        //     "firewallMode": "Detection",
+        //     "ruleSetType": "OWASP",
+        //     "ruleSetVersion": "3.0",
+        //     "disabledRuleGroups": [],
+        //     "requestBodyCheck": true,
+        //     "maxRequestBodySizeInKb": 128,
+        //     "fileUploadLimitInMb": 100
+        // },
+        // "enableHttp2": true,
```
Comment (eriqua) [inline arm/Microsoft.Network/applicationGateways/.parameters/parameters.json:23]: Is this to be removed?

Comment (rahalan) [reply arm/Microsoft.Network/applicationGateways/.parameters/parameters.json:23]: Unfortunately not. It is only sometimes working. Need to iterate on that

File: arm/Microsoft.Network/applicationGateways/deploy.bicep [hunk 4]
```diff
@@ -151,244 +240,108 @@ var diagnosticsMetrics = [for metric in metricsToEnable: {
...
+    //     }
+    //   ]
+    //   fileUploadLimitInMb: int
+    //   firewallMode: 'string'
+    //   maxRequestBodySize: int
+    //   maxRequestBodySizeInKb: int
+    //   requestBodyCheck: bool
+    //   ruleSetType: 'string'
+    //   ruleSetVersion: 'string'
+    // }
```
Comment (eriqua) [inline arm/Microsoft.Network/applicationGateways/deploy.bicep:340]: Is this to be removed?

File: arm/Microsoft.Network/applicationGateways/.parameters/parameters.json [hunk 5]
```diff
@@ -5,157 +5,356 @@
...
-                            "ipAddress": "10.0.0.4"
+                    "name": "privateVmHttpSetting",
+                    "properties": {
+                        "port": 80,
+                        "protocol": "Http",
+                        "cookieBasedAffinity": "Disabled",
+                        "pickHostNameFromBackendAddress": false,
+                        "requestTimeout": 30,
+                        "probe": {
+                            "id": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/applicationGateways/sxx-az-apgw-x-001/probes/privateVmHttpSettingProbe"
```
Comment (eriqua) [inline arm/Microsoft.Network/applicationGateways/.parameters/parameters.json:69]: `sxx` to be replaced with `<<namePrefix>>` in all its occurrences

File: arm/Microsoft.Network/applicationGateways/deploy.bicep (hunk 4 above)
Comment (rahalan) [reply arm/Microsoft.Network/applicationGateways/deploy.bicep:340]: Ooopsie. Have to look into that one. Sorry

File: arm/Microsoft.Network/applicationGateways/deploy.bicep [hunk 6]
```diff
@@ -18,100 +88,112 @@ param sku string = 'WAF_Medium'
...
 
-@description('Optional. SSL certificate reference name for a certificate stored in the Key Vault to configure the HTTPS listeners.')
-param sslCertificateName string = 'sslCertificate01'
+// @description('Optional. Ssl protocols to be disabled on application gateway.')
+// @allowed([
+//   'TLSv1_0'
+//   'TLSv1_1'
+//   'TLSv1_2'
+// ])
+// param sslPolicyDisabledSslProtocols array = []
```
Comment (eriqua) [inline arm/Microsoft.Network/applicationGateways/deploy.bicep:136]: Is this to be removed?

File: arm/Microsoft.Network/applicationGateways/deploy.bicep [hunk 7]
```diff
@@ -151,244 +240,108 @@ var diagnosticsMetrics = [for metric in metricsToEnable: {
...
+    rewriteRuleSets: rewriteRuleSets
+    sku: {
+      name: sku
+      tier: endsWith(sku, 'v2') ? sku : substring(sku, 0, indexOf(sku, '_'))
+      capacity: autoscaleMaxCapacity > 0 && autoscaleMinCapacity > 0 ? null : capacity
+    }
+    sslCertificates: sslCertificates
+    sslPolicy: {
+      cipherSuites: sslPolicyCipherSuites
+      // disabledSslProtocols: sslPolicyDisabledSslProtocols
```
Comment (eriqua) [inline arm/Microsoft.Network/applicationGateways/deploy.bicep:306]: Is this to be removed?

Comment (rahalan) [reply arm/Microsoft.Network/applicationGateways/deploy.bicep:306]: Yes. Not supported anymore. Will remove it.

**Review by eriqua — COMMENTED** [review]
I'd suggest attaching the status badge to the PR

File: arm/Microsoft.Network/applicationGateways/.parameters/parameters.json (hunk 3 above)
Comment (rahalan) [reply arm/Microsoft.Network/applicationGateways/.parameters/parameters.json]: fixed it

File: arm/Microsoft.Network/applicationGateways/deploy.bicep (hunk 4 above)
Comment (rahalan) [reply arm/Microsoft.Network/applicationGateways/deploy.bicep]: fixed it

File: arm/Microsoft.Network/applicationGateways/.parameters/parameters.json (hunk 5 above)
Comment (rahalan) [reply arm/Microsoft.Network/applicationGateways/.parameters/parameters.json]: done

File: arm/Microsoft.Network/applicationGateways/deploy.bicep (hunk 6 above)
Comment (rahalan) [reply arm/Microsoft.Network/applicationGateways/deploy.bicep]: fixed it

File: arm/Microsoft.Network/applicationGateways/deploy.bicep (hunk 7 above)
Comment (rahalan) [reply arm/Microsoft.Network/applicationGateways/deploy.bicep]: done

**Comment by rahalan** [comment]
> I'd suggest attaching the status badge to the PR

[Network: ApplicationGateways](https://github.com/Azure/ResourceModules/actions/workflows/ms.network.applicationgateways.yml)

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**